src/mlpack/core/util/mlpack_main.hpp

```diff
@@ -92,26 +93,32 @@ static const std::string testName = "";
 #include <mlpack/bindings/cli/parse_command_line.hpp>
 #include <mlpack/bindings/cli/end_program.hpp>
 
-static void mlpackMain(); // This is typically defined after this include.
+#ifndef BINDING_NAME
+  #error "BINDING_NAME not defined!"
+#endif
+
+static void BINDING_NAME(mlpack::util::Params&, mlpack::util::Timers&);
 
 int main(int argc, char** argv)
```

Comment:
Here are the changes to how `main()` works in the command-line bindings now.  You can see instead of calling `mlpackMain()`, we call `BINDING_NAME()`.  This also means we can have more than one binding defined in a single translation unit---previously, we required that each binding was in its own translation unit because they were all called `mlpackMain()`.

This also means that we could now generate *one* `.so` file for *all* bindings (though it might be large), instead of one `.so` file for *each* binding.  We can apply this same set of changes for any language, including R and Python, and I think we will also see compilation time improvements for the bindings (which is probably something @eddelbuettel will rejoice about :smile:).

Reply:
A single `.so` would be nice to have. One problem I think (I am not sure) that can arise is that whenever I do some changes to a single binding then will it rebuild all the other bindings also?

Reply:
Actually CMake should be able to avoid doing that---while we are combining each binding's `_main.cpp` into the final `.so`, each individual file will produce its own `.o` file, which can be cached for future compilations.  So if you only modify one, then that one binding's `.o` will be recompiled, and then combined with all the other `.o`s whose compilation is cached.  At least, I *think* that's what will happen. :smile: